### PR TITLE
generate-zephyr-dts.py: refactored version

### PR DESCRIPTION
**Pull request by hvegh**
The changes are part of an effort to enable platformio to handle the
different litex-board variants for the zephyr framework.

- Removed the dependency on the intermediate interpretations from
  the configuration class. No external python modules required,
  can run as standalone conversion tool.
- Less code with similar output results.
- Switched to JSON as it is better in preserving structure than CSV
  format.
- Refactored the formatting and translation handlers.

The resulting changes are aimed for better maintainability.

From a maintenance perspective it would be nice if this file could
be part of the zephyr mainline tree.

Signed-off-by: Henk Vergonet <henk.vergonet@gmail.com>

### CODE REVIEW

**Comment by mithro** [comment]
@hvegh -- Looks like this change needs to be rebased.

**Comment by mithro** [comment]
FYI - @mateusz-holenko

File: generate-zephyr-dts.py [hunk 1]
```diff
@@ -1,179 +1,136 @@
...
-    result = ""
+def generate_dts_config(csr):
+    dts = cnf = ''
 
-    for name, peripheral in configuration.peripherals.items():
-        if name not in peripheral_handlers:
-            print('Skipping unsupported peripheral `{}` at {}'
-                  .format(name, hex(peripheral['address'])))
-            continue
+    for name, parm in overlay_handlers.items():
```
Comment (mateusz-holenko) [inline generate-zephyr-dts.py]: By iterating over `overlay_handler` instead of `configuration.peripherals` you miss the warnings about unsupported (yet available in the platform) peripherals.
I think this might be important when debugging problems.

File: generate-zephyr-dts.py [hunk 2]
```diff
@@ -1,179 +1,136 @@
...
     'ethmac': {
         'handler': ethmac_handler,
-        'reference': 'eth0',
-        'size': 0x6c,
-        'buffer': lambda: configuration.mem_regions['ethmac'],
+        'alias': 'eth0',
+        'size': 0x80,
         'config_entry': 'ETH_LITEETH'
     },
+    'spiflash': {
```
Comment (mateusz-holenko) [inline generate-zephyr-dts.py]: Support for new peripheral should be added in a separate commit.

File: generate-zephyr-dts.py [hunk 3]
```diff
@@ -1,179 +1,136 @@
...
-
-
-peripheral_handlers = {
+overlay_handlers = {
     'uart': {
         'handler': peripheral_handler,
-        'reference': 'uart0',
-        'size': 0x18,
+        'alias': 'uart0',
+        'size': 0x20,
```
Comment (mateusz-holenko) [inline generate-zephyr-dts.py]: Was the previous `size` value wrong? If so I suggest to extract this change into a separate commit.

**Review by mateusz-holenko — COMMENTED** [review]
Hi @hvegh,

Thanks for your contribution!

I like the proposed way of generating dtsi - I agree it's cleaner and easier to follow.

I'm not sure about the `json` vs `csv`, though. Currently we use the "external" script for parsing because:
* it supports both formats (`json` and `csv`),
* this is a common logic of `generate-zephyr-dts.py` and `generate-renode-scripts.py`.

Of course I understand that having a single file might be more convenient, but is this necessary? I'm open for a discussion.

File: generate-zephyr-dts.py (hunk 1 above)
Comment (hvegh) [reply generate-zephyr-dts.py]: Hi @mateusz-holenko,

Thanks you for the review!

Yes I can see your point, the shared code between the DTS and Renode configuration.

Currently there are 6 parsers or interpretations made:
`LiteX to CSV  ---  CSV to lx.configutation`
`LiteX to JSON  ---  JSON to lx.configutation`
`lx.configuration to renode`
`lx.configutation to DTS`

I would argue to at least ditch the `CSV` format in favor of `JSON` as it is better at preserving structure and type. So `JSON` seems more future proof, and we can ditch `CSV` at zero cost, nobody gets hurts as far as i can tell..
So then we have to maintain two parsers less:

`LiteX to JSON  ---  JSON to lx.configutation`
`lx.configuration to renode`
`lx.configutation to DTS`

Similarl since the `CSV` is a subset of `JSON` and there are no complex routines that can be shared, for now. The benefits of having a separate configuration class just for abstracting the JSON and CSV inputs is not that usefull. So you could further reduce to:

`LiteX to JSON`
`JSON to renode`
`JSON to DTS`

We end up with two parsers closely coupled to the JSON output format of LITEX.

Comment (hvegh) [reply generate-zephyr-dts.py]: > By iterating over `overlay_handler` instead of `configuration.peripherals` you miss the warnings about unsupported (yet available in the platform) peripherals.
> I think this might be important when debugging problems.

I agree this should be fixed!

File: generate-zephyr-dts.py (hunk 2 above)
Comment (hvegh) [reply generate-zephyr-dts.py]: SPI is tested and working.

No problem will split.

File: generate-zephyr-dts.py (hunk 3 above)
Comment (hvegh) [reply generate-zephyr-dts.py]: Just looking at the generated configuration from litex:

Last uart addres is:
`uart_rxfull  CSR_BASE + 0x1c`

So I expanded the IO region to include this register.

File: generate-zephyr-dts.py (hunk 1 above)
Comment (mateusz-holenko) [reply generate-zephyr-dts.py]: > we can ditch CSV at zero cost, nobody gets hurts as far as i can tell

Do you mean that we can parse CSV files using the json parser?  litex-buildenv uses this script to generate Zephyr overlay from CSV: https://github.com/timvideos/litex-buildenv/blob/master/scripts/build-zephyr.sh#L84.

**Comment by mateusz-holenko** [comment]
BTW there is no need to close the PR and open a new one - you can always force push the original branch and GitHub will handle this nicely.

File: generate-zephyr-dts.py (hunk 1 above)
Comment (mateusz-holenko) [reply generate-zephyr-dts.py]: Additionally https://github.com/google/CFU-Playground also uses the csv format for generating Renode scripts at the moment.

Comment (hvegh) [reply generate-zephyr-dts.py]: Nice links, both are interesting projects. 
The main drawback for JSON is that the addresses are in decimal instead of hex, so it is not as human readable as CSV.
